docs/pages/api/automatically-register-agents.mdx

```diff
@@ -0,0 +1,988 @@
+
+```go
+func (t *tokenDemoApp) listAppContainerURLs(ctx context.Context, image string) (map[string]struct{}, error) {
+	c, err := t.dockerClient.ContainerList(ctx, dtypes.ContainerListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "ancestor",
+			Value: image,
+		}),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	l := make(map[string]struct{})
+
+	for _, r := range c {
+		b, ok := r.NetworkSettings.Networks[networkName]
+		// Not connected to the chosen network, so skip it
+		if !ok {
+			continue
+		}
+
+		u, err := url.Parse(fmt.Sprintf(
+			"http://%v:%v",
+			b.IPAddress,
+			managementPort,
+		))
+
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		l[u.String()] = struct{}{}
+	}
+
+	return l, nil
+}
+```
+
+This function uses the `tokenDemoApp.dockerClient` field, the Docker API client
+library's `Client` type, to send the Docker daemon a request to list containers.
+The `dtypes.ContainerListOptions` struct instructs the Docker daemon to list
+only the containers that have the image we specify in the `image` parameter.
+
+For each container returned by the Docker daemon, we look up the container's IP
+address within a predetermined network (the default bridge network). We know in
+advance that all RabbitMQ containers will have the management port opened, so we
+use the container's IP address and the management port to compose a URL and
+insert it into the map that we will return from this function.
+
+Note that the map this function returns assigns URL strings to empty structs.
+In Go, the empty struct consumes no memory. Go programs often use the empty
+struct for the values of hash tables, since a program can search for a key in
+constant time without using the value.
+
+### Create a token for a new Application Service instance
+
+After reconciling the map of registered applications with the map of application
+containers, we will need to:
+
+- Launch Application Service instances to proxy unregistered application
+  containers.
+- Delete Application Service instances that no longer correspond to a running
+  application.
+
+To launch a new Application Service instance, we will create a **join token**.
+Teleport's services can join a cluster by presenting a token to the Auth
+Service.
+
+Add the following two functions below `listAppContainerURLs`:
+
+```go
+func cryptoRandomHex(len int) (string, error) {
+	randomBytes := make([]byte, len)
+	if _, err := rand.Reader.Read(randomBytes); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return hex.EncodeToString(randomBytes), nil
+}
+
+func (t *tokenDemoApp) createAppToken(ctx context.Context) (string, error) {
+	n, err := cryptoRandomHex(tokenLenBytes)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	tok, err := types.NewProvisionTokenFromSpec(
+		n,
+		time.Now().Add(tokenTTL),
+		types.ProvisionTokenSpecV2{
+			Roles: types.SystemRoles{types.RoleApp},
+		})
+
+	if err := t.teleportClient.CreateToken(ctx, tok); err != nil {
```

Comment:
This example might be simpler if you registered dynamic apps rather than started new app services and generated new join tokens.

Reply:
I wanted to demonstrate the token method in this guide somehow so users could see an example, but it makes sense to advise using a dynamic approach for simplicity. I've moved the `Details` box about the dynamic approach higher in this section and changed the first paragraph after the code snippet to make the dynamic approach more prominent.